src/coreComponents/physicsSolvers/multiphysics/CompositionalMultiphaseReservoirAndWells.cpp

```diff
@@ -0,0 +1,406 @@
+template<> class CompositionalCatalogNames< CompositionalMultiphaseBase >
+{
+public:
+  static string name() { return "CompositionalMultiphaseReservoir"; }
+};
+/*
+   // Class specialization for a RESERVOIR_SOLVER set to MultiphasePoromechanics
+   template<> class CompositionalCatalogNames< MultiphasePoromechanics >
+   {
+   public:
+   static string name() { return "CompositionalMultiphasePoromechanicsReservoir"; }
+   };
+ */
+
+}
+
+// provide a definition for catalogName()
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+string
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+catalogName()
+{
+  return CompositionalCatalogNames< COMPOSITIONAL_RESERVOIR_SOLVER >::name();
+}
+
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+CompositionalMultiphaseReservoirAndWells( const string & name,
+                                          Group * const parent )
+  : Base( name, parent )
+{}
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+~CompositionalMultiphaseReservoirAndWells()
+{}
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+void
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+postProcessInput()
+{
+  Base::postProcessInput();
+
+  integer const & useMassFlow = Base::getReservoirSolver()->template getReference< integer >( CompositionalMultiphaseBase::viewKeyStruct::useMassFlagString() );
+  integer const & useMassWell = Base::getWellSolver()->template getReference< integer >( CompositionalMultiphaseWell::viewKeyStruct::useMassFlagString() );
+  GEOSX_THROW_IF( useMassFlow != useMassWell,
+                  GEOSX_FMT( "CompositionalMultiphaseReservoir '{}': the input flag {} must be the same in the flow and well solvers, respectively '{}' and '{}'",
+                             this->getName(), CompositionalMultiphaseBase::viewKeyStruct::useMassFlagString(),
+                             Base::getReservoirSolver()->getName(), Base::getWellSolver()->getName() ),
+                  InputError );
+}
+
+template< typename COMPOSITIONAL_RESERVOIR_SOLVER >
+void
+CompositionalMultiphaseReservoirAndWells< COMPOSITIONAL_RESERVOIR_SOLVER >::
+initializePreSubGroups()
+{
+  if( catalogName() == CompositionalCatalogNames< CompositionalMultiphaseBase >::name() )
+  {
+    if( dynamicCast< CompositionalMultiphaseFVM * >( this->getReservoirSolver() ) )
+    {
+      m_linearSolverParameters.get().mgr.strategy = LinearSolverParameters::MGR::StrategyType::compositionalMultiphaseReservoirFVM;
+    }
+    else if( dynamicCast< CompositionalMultiphaseHybridFVM * >( this->getReservoirSolver() ) )
+    {
+      m_linearSolverParameters.get().mgr.strategy = LinearSolverParameters::MGR::StrategyType::compositionalMultiphaseReservoirHybridFVM;
+    }
```

Comment:
I wonder if instead of dynamic casting we could just switch on the child solver's MGR strategy type, i.e.
```c++
switch( this->getReservoirSolver()->getLinearSolverParameters().mgr.strategy )
{
  case LinearSolverParameters::MGR::StrategyType::compositionalMultiphaseFVM: { ...; break; }
  case LinearSolverParameters::MGR::StrategyType::compositionalMultiphaseHybridFVM { ...; break; }
}
```
This still contains a large degree of hardcoding to specific solver types, but enables us to not include `CompositionalMultiphaseFVM.hpp`, etc. at all in this file (in combination with my other suggestion about catalog names, which may or may not be implemented in this PR).

The only issue might be that for `SinglePhaseFVM` there is currently no MGR strategy, but a trivial single-level one could be added for consistency. Again, this can be addressed at a later point.

Reply:
Yes, I agree with the suggestion, I will implement the change before merging